shift_proposal.h

```diff
@@ -2,33 +2,128 @@
+
+template<class I>
+using iterator_category_t = typename std::iterator_traits<I>::iterator_category;
+
+template<class I, class Tag, class = void>
+constexpr bool is_category = false;
+template<class I, class Tag>
+constexpr bool is_category<I, Tag, std::enable_if_t<
+    std::is_convertible_v<iterator_category_t<I>, Tag>>> = true;
+
+/// Increment (decrement for negative n) i |n| times or until i == bound,
+/// whichever comes first. Returns n + the difference between i's final position
+/// and its initial position. (Note: "advance" has overloads with this behavior
+/// in the Ranges TS.)
+template<class I>
+constexpr difference_type_t<I> bounded_advance(
+    I& i, difference_type_t<I> n, I const bound)
+{
+    if constexpr (is_category<I, std::bidirectional_iterator_tag>) {
+        for (; n < 0 && i != bound; ++n, void(--i)) {
+            ;
+        }
+    }
+
+    for(; n > 0 && i != bound; --n, void(++i)) {
+        ;
+    }
+
+    return n;
+}
+
 template<class ForwardIt>
-ForwardIt shift_left(ForwardIt first, ForwardIt last, typename std::iterator_traits<ForwardIt>::difference_type n, std::optional<typename std::iterator_traits<ForwardIt>::value_type> filler = std::nullopt)
+ForwardIt shift_left(ForwardIt first, ForwardIt last, difference_type_t<ForwardIt> n)
 {
-    assert(n >= 0);
-    if (n == 0)
+    if (n <= 0) {
         return last;
-    auto shifted_last = std::move(std::next(first, n), last, first);
-    if (filler.has_value())
-        std::fill(shifted_last, last, *filler);
-    return shifted_last;
+    }
+
+    auto mid = first;
+    if (::bounded_advance(mid, n, last)) {
+        return first;
+    }
+
+    return std::move(std::move(mid), std::move(last), std::move(first));
 }
 
-template<class BidirIt>
-BidirIt shift_right(BidirIt first, BidirIt last, typename std::iterator_traits<BidirIt>::difference_type n, std::optional<typename std::iterator_traits<BidirIt>::value_type> filler = std::nullopt)
+template<class ForwardIt>
+ForwardIt shift_right(ForwardIt first, ForwardIt last, difference_type_t<ForwardIt> n)
 {
-    assert(n >= 0);
-    if (n == 0)
+    if (n <= 0) {
         return first;
-    auto shifted_first = std::move_backward(first, std::prev(last, n), last);
-    if (filler.has_value())
-        std::fill(first, shifted_first, *filler);
-    return shifted_first;
+    }
+
+    if constexpr (is_category<ForwardIt, std::bidirectional_iterator_tag>) {
+        auto mid = last;
+        if (::bounded_advance(mid, -n, first)) {
+            return last;
+        }
+        return std::move_backward(std::move(first), std::move(mid), std::move(last));
+    } else {
+        auto result = first;
+        if (::bounded_advance(result, n, last)) {
+            return last;
+        }
+
+        // Invariant: next(first, n) == result
+        // Invariant: next(trail, n) == lead
+
+        auto lead = result;
+        auto trail = first;
+
+        for (; trail != result; ++lead, void(++trail)) {
```

Comment:
Overload operator comma paranoia. Users aren't forbidden to overload operator comma on the types that pass to the standard library, so implementations need to defend against such shenanigans. (`&` overloads are the other big offendor.)